source/cbor.c

```diff
@@ -0,0 +1,708 @@
+    struct aws_byte_buf encoded_buf;
+};
+
+struct aws_cbor_encoder *aws_cbor_encoder_new(struct aws_allocator *allocator, size_t init_buffer_size) {
+    struct aws_cbor_encoder *encoder = aws_mem_acquire(allocator, sizeof(struct aws_cbor_encoder));
+    encoder->allocator = allocator;
+    aws_byte_buf_init(&encoder->encoded_buf, allocator, init_buffer_size);
+
+    /* TODO: refcount */
+    return encoder;
+}
+
+struct aws_cbor_encoder *aws_cbor_encoder_release(struct aws_cbor_encoder *encoder) {
+    aws_byte_buf_clean_up(&encoder->encoded_buf);
+    aws_mem_release(encoder->allocator, encoder);
+    return NULL;
+}
+
+struct aws_byte_cursor aws_cbor_encoder_get_encoded_data(struct aws_cbor_encoder *encoder) {
+    return aws_byte_cursor_from_buf(&encoder->encoded_buf);
+}
+
+void aws_cbor_encoder_reset_encoded_data(struct aws_cbor_encoder *encoder) {
+    aws_byte_buf_reset(&encoder->encoded_buf, false);
+}
+
+#define ENCODE_THROUGH_LIBCBOR(length_to_reserve, encoder, value, fn)                                                  \
+    do {                                                                                                               \
+        aws_byte_buf_reserve_dynamic(&(encoder)->encoded_buf, (encoder)->encoded_buf.len + (length_to_reserve));       \
+        size_t encoded_len =                                                                                           \
+            fn(value,                                                                                                  \
+               (encoder)->encoded_buf.buffer + (encoder)->encoded_buf.len,                                             \
+               (encoder)->encoded_buf.capacity - (encoder)->encoded_buf.len);                                          \
+        AWS_ASSERT((encoded_len) != 0);                                                                                \
+        (encoder)->encoded_buf.len += (encoded_len);                                                                   \
+    } while (false)
+
+void aws_cbor_encode_uint(struct aws_cbor_encoder *encoder, uint64_t value) {
+    ENCODE_THROUGH_LIBCBOR(9, encoder, value, cbor_encode_uint);
+}
+
+void aws_cbor_encode_negint(struct aws_cbor_encoder *encoder, uint64_t value) {
+    ENCODE_THROUGH_LIBCBOR(9, encoder, value, cbor_encode_negint);
+}
+
+void aws_cbor_encode_single_float(struct aws_cbor_encoder *encoder, float value) {
+    ENCODE_THROUGH_LIBCBOR(5, encoder, value, cbor_encode_single);
+}
+
+void aws_cbor_encode_double(struct aws_cbor_encoder *encoder, double value) {
+    if (isnan(value) || isinf(value)) {
+        /* For special value: NAN/INFINITY, type cast to float and encode into single float. */
+        aws_cbor_encode_single_float(encoder, (float)value);
+        return;
+    }
+    /* Conversation from int to floating-type is implementation defined if loss of precision */
+    if (value <= (double)INT64_MAX && value >= (double)INT64_MIN) {
```

Comment:
Just want to simplify the implementation. 

I guess in real life if we cast double to 64bit int, we don't actually have benefit to save any bytes here.

Reply:
im fine on losing optimization opportunity here since its a corner case that is not likely to occur in practice, but we should doc that it was done in order to simplify code.